dot/sync/chain_sync_test.go

```diff
@@ -0,0 +1,1196 @@
+	peersTried[testPeerA] = struct{}{}
+	req.StartingBlock = variadic.Uint32OrHash{}
+	peers = cs.determineSyncPeers(req, peersTried)
+	require.Equal(t, 1, len(peers))
+	require.Equal(t, []peer.ID{testPeerB}, peers)
+}
+
+func Test_chainSync_logSyncSpeed(t *testing.T) {
+	t.Parallel()
+
+	ctrl := gomock.NewController(t)
+
+	mockBlockState := NewMockBlockState(ctrl)
+	mockBlockState.EXPECT().BestBlockHeader().Return(&types.Header{
+		Number: 1,
+	}, nil).AnyTimes()
+	mockBlockState.EXPECT().GetHighestFinalisedHeader().Return(&types.Header{
+		Number: 1,
+	}, nil).AnyTimes()
+
+	mockNetwork := NewMockNetwork(ctrl)
+	mockNetwork.EXPECT().Peers().Return([]common.PeerInfo{}).AnyTimes()
+
+	type fields struct {
+		blockState  BlockState
+		network     Network
+		state       chainSyncState
+		benchmarker *syncBenchmarker
+	}
+	tests := []struct {
+		name   string
+		fields fields
+	}{
+		{
+			name: "state bootstrap",
+			fields: fields{
+				blockState:  mockBlockState,
+				network:     mockNetwork,
+				benchmarker: newSyncBenchmarker(10),
+				state:       bootstrap,
+			},
+		},
+		{
+			name: "case tip",
+			fields: fields{
+				blockState:  mockBlockState,
+				network:     mockNetwork,
+				benchmarker: newSyncBenchmarker(10),
+				state:       tip,
+			},
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			ctx, cancel := context.WithCancel(context.Background())
+			cs := &chainSync{
+				ctx:           ctx,
+				cancel:        cancel,
+				blockState:    tt.fields.blockState,
+				network:       tt.fields.network,
+				state:         tt.fields.state,
+				benchmarker:   tt.fields.benchmarker,
+				logSyncPeriod: time.Nanosecond,
+			}
+			go cs.logSyncSpeed()
+			time.Sleep(time.Nanosecond)
+			cancel()
+		})
+	}
+}
+
+func newMockBlockStateForChainSyncTests(ctrl *gomock.Controller) BlockState {
+	mock := NewMockBlockState(ctrl)
+	mock.EXPECT().BestBlockHeader().Return(&types.Header{}, nil).AnyTimes()
+	mock.EXPECT().HasHeader(gomock.AssignableToTypeOf(common.Hash{})).Return(true, nil).AnyTimes()
+	return mock
+}
+
+func newMockDisjointBlockSet(ctrl *gomock.Controller) DisjointBlockSet {
+	mock := NewMockDisjointBlockSet(ctrl)
+	mock.EXPECT().run(gomock.Any()).AnyTimes()
+	mock.EXPECT().addHeader(gomock.AssignableToTypeOf(&types.Header{})).Return(nil).AnyTimes()
+	mock.EXPECT().addHashAndNumber(gomock.AssignableToTypeOf(common.Hash{}),
+		gomock.AssignableToTypeOf(uint(0))).Return(nil).AnyTimes()
+	return mock
+}
+
+func Test_chainSync_start(t *testing.T) {
+	t.Parallel()
+
+	ctrl := gomock.NewController(t)
+	type fields struct {
+		blockState    BlockState
+		pendingBlocks DisjointBlockSet
+		benchmarker   *syncBenchmarker
+		slotDuration  time.Duration
+	}
+	tests := []struct {
+		name   string
+		fields fields
+	}{
+		{
+			name: "base case",
+			fields: fields{
+				blockState:    newMockBlockStateForChainSyncTests(ctrl),
+				pendingBlocks: newMockDisjointBlockSet(ctrl),
```

Comment:
configure these by passing them the subtest controller instead of the parent test controller